### ISSUE TITLE
Allow for annotation to replace empty no-args constructor with Inject

### ISSUE THREAD

**Issue by Addvilz** [google/dagger#1411]
There has been a lot of discussion about the no-args `@Inject` constructors already. To that end, I would like to propose an alternative solution to the boilerplate-y looking constructor - an annotation that would explicitly mark the class as injectable by Dagger, even without `@Inject` marked constructor. Both options could live on, for those who prefer one or another.

Something like this:

```java
import dagger.Injectable;

@Injectable
public class MyService {
    // ...
}
```

If the annotation is present, Dagger would allow for "default" no-args constructor to be used.

Related to #1105, #1409 and discussed in square#412.

**Comment by ronshapiro** [comment]
I don't think adding support for something that is so similar to an already standard solution makes sense. Then readers have to discern why one solution was used and not the other. As it is, `@Inject MyService() {}` is just as easy to write, plus it will work with all other JSR 330 injectors.

**Comment by Addvilz** [comment]
I understand that, however, dangling empty constructors are not exactly a good solution to this problem either.

Isn't it the case that JSR-330 has no requirement to have empty constructor and other frameworks do not enforce this either? Not sure exactly how this would conflict with other dependency injection frameworks, given that the custom annotation would be ignored by them and "default" constructor would be used. 

To avoid confusion, `@Injectable` should not be allowed for classes with defined constructors.

**Comment by ronshapiro** [comment]
You're right that JSR 330 does allow default constructors to be injectable, but it turns out that this is very commonly quite confusing to users ("I never bound this, where is it coming from?") and it is the opposite of being declarative. We've discussed this elsewhere on the issue tracker before, and we're not interested in changing that part of Dagger.

I wouldn't use the same language that you do for "dangling empty constructor". It's not dangling - it is explicitly documenting an intention.

**Comment by Addvilz** [comment]
I have to disagree with that. The use of otherwise empty code block (with an annotation, might I add) just for documentation purposes seems counter-intuitive and even superfluous at times - plenty of times I have had to explain someone unfamiliar with Dagger that yes, this is required, and the reason for having it. I believe an annotation would much more natural choice being that the purpose of annotations is to hold metadata like this.

Bottom line being, I am not asking to change this "explicit" behavior of Dagger. Frankly, I agree with the concept completely. I am just asking you to consider alternative means of achieving the same goal in a way that seems more native to the language. Not as a replacement even, but as an optional extra.

**Comment by ronshapiro** [comment]
Each new feature adds maintenance cost and conceptual overhead. Some consider it flexibility, others consider it confusing. We've opted for fewer concepts unless there's a clear benefit (runtime, compile time static analysis options, etc). I don't think this has much benefits besides some personal style.

If we allow custom binding annotations at some point, this would be a great candidate for something that you could add to your project (but keep outside of Dagger).
